Api/endpoints

kennankole opened the pull request:
### Creating  BlogApp Api Endpoints
* Created an API endpoint to list all posts for a user.
* Created an API endpoint to list all comments for a user's post.
* Created an API endpoint to add a comment to a post

kennankole committed:
Eager loaded posts: Remove N+1 problem from the post view

kennankole committed:
Installed Capybara

kennankole committed:
Written intergration tests for users

kennankole committed:
Written intergration tests for users#show

kennankole committed:
Written intergration test for posts#index

kennankole committed:
Written intergration test for posts#show

kennankole committed:
Refactored unit tests

kennankole committed:
Fixing linter errors

kennankole committed:
Setting up devise for authentication

kennankole committed:
Updating the application_controller

kennankole committed:
Added email column to the users table

kennankole committed:
Generated custom views from devise

kennankole committed:
Added an name field to the sign up form

kennankole committed:
Implemented basic authentication with devise
Users can log in using email
Added ask for confirmation for users
Verified that users can reset passwords

kennankole committed:
Installed tailwind css

kennankole committed:
Styled user registration and login forms

kennankole committed:
Removed tailwind
Fixed all linter errors

kennankole committed:
Fixing linter error

kennankole committed:
Fixing linter errors
Changed sign up header htm tag

kennankole committed:
Installed CanCanCan
Added role column to the users table

kennankole committed:
Implemented deleting post by an authorized user

kennankole committed:
Implemented comment delete operation

kennankole committed:
Fixing linter errors

kennankole committed:
Created an API endpoint to list all posts for a user.

kennankole committed:
Created an API endpoint to list all comments for a user's post.

kennankole committed:
Created an API endpoint to add comment to a post

kennankole committed:
Fixing linter errors

BertrandConxy commented on config/routes.rb:
```diff
@@ -1,23 +1,49 @@
+        end
+      end
+    end
+  end
```
- [ ] Good job so far 👏 👏 However from the coding best practice perspective, there is much repetition of codes here in your routes. The API endpoint routes can be grouped in one namespace so as to avoid Repeating Yourself (DRY principle). 
for simplicity, you can make it like 👇 
```ruby
namespace :api do
 namespace :v1 do
   resources :users do
      resources :posts do
         resources :comments
       end
   end
 end
end
```
I hope you find this helpful 😃

BertrandConxy reviewed (changes requested):
Hi @kennankole, 

Good job so far! 👏 👏 
There are some issues that you still need to work on to go to the next project but you are almost there! 🏃‍♂️ 🏃‍♂️ 


## Highlights 💪 
- No linter errors
- correct flow is used
- good PR


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @BertrandConxy in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

kennankole committed:
Refactored routes to remove duplicate routes to the API
Changed
from
devise_for :users, controllers: { confirmations: 'users/confirmations' }
to
to devise_for :users, controllers: { confirmations: 'confirmations' }

kennankole committed:
Refactored the routes api routes

MeqdamAlqudah commented on config/routes.rb:
```diff
@@ -1,23 +1,34 @@
   resources :users, only: [:index, :show, :create] do
-    resources :posts, only: [:new, :show, :index, :create] do
-      resources :comments, only: [:new, :create]
+    resources :posts, only: [:new, :show, :index, :create, :destroy] do
+      resources :comments, only: [:new, :create, :destroy]
       member do
         put "like", to: "posts#like"
         delete "unlike", to: "likes#destroy"
       end
     end
   end
 
-  resources :posts, only: [] do
+  resources :posts, only: [:new, :show, :index, :create, :destroy] do
     member do
       get "like", to: "posts#like", as: :like_post
       get "unlike", to: "posts#unlike", as: :unlike_post
     end
   end
+  delete "posts/:id/remove", to: "posts#destroy", as: "remove_post"
+  delete "users/:id/posts/:id/comments/:id/remove", to: "comments#destroy", as: "remove_comment"
+
+  namespace :api do
+    namespace :v1 do
+      resources :users, only: [:index] do
+        resources :posts, only: [:index, :show] do
+          resources :comments, only: [:index, :create]
+        end
+        resources :comments, only: [:index]
+      end
+    end
+  end
```
- [ Optional] Please it would be better to apply the dry principle, you can just use the code from lines 24 to 33 to set your routes there is no need from repeating the same routes also make sure to add all the methods you used in posts, likes and comments and users after this.

MeqdamAlqudah reviewed (approved):
Hi @kennankole , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
![image](https://user-images.githubusercontent.com/74744787/224358022-7bf57996-a876-4267-93c2-7e6ac810c2c8.png)

## Highlights 
:+1: no linter errors
:+1: the app is working

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
